qringbuf.go

```diff
@@ -1,3 +1,191 @@
+//    ◦ NextRegion(…) has not been called again allowing further writes into the buffer
+//    ◦ *Region.Reserve() was invoked, which blocks writes until a *Region.Release()
+//
+// Examples
+//
+// In code the basic usage looks roughly like this:
+//
+//  qrb, initErr := qringbuf.NewFromReader( *someReader, qringbuf.Config{ … } )
+//  …
+//  readLimit = int64( stopReadingAfterThatManyBytes )
+//  startErr := qrb.StartFill( readLimit )
+//  …
+//  var available, processed int
+//  for {
+//      // Reevaluate available and processed from the *previous* round,
+//      // indicating how much has NOT been processed, and needs re-serving
+//      reg, streamErr := qrb.NextRegion( available - processed )
+//
+//      // io.EOF only means the collector stopped: there could be up to
+//      // BufferSize bytes remaining in the buffer. Loop until reg is nil
+//      if ( streamErr != nil && streamErr != io.EOF ) || reg == nil {
+//          return streamErr
+//      }
+//
+//      // Work with region, processing all or just a portion of the data
+//      available = reg.Size()
+//      processed = frobnicate( reg.Bytes(), … )
+//  }
+//
+// In addition one can operate over individual (sub)regions with "fearless
+// concurrency":
+//
+//  …
+//  var available, processed int
+//  for {
+//      // Reevaluate available and processed from the *previous* round,
+//      // indicating how much has NOT been processed, and needs re-serving
+//      reg, streamErr := qrb.NextRegion( available - processed )
+//
+//      // io.EOF only means the collector stopped: there could be up to
+//      // BufferSize bytes remaining in the buffer. Loop until reg is nil
+//      if ( streamErr != nil && streamErr != io.EOF ) || reg == nil {
+//          return streamErr
+//      }
+//
+//      available = reg.Size()
+//      if available > 256 {
+//          reg = reg.SubRegion( 0, 256 )
+//      }
+//
+//      reg.Reserve()
+//      processed = reg.Size()
+//
+//      go func() {
+//          frobnicate( reg.Bytes(), … )
+//          reg.Release()
+//      }()
+//  }
+//
+// In all cases the background collector goroutine reading from *someReader
+// into the ring buffer is guaranteed to:
+//  - never overwrite the buffer portion backing the latest result of NextRegion(…)
+//  - never overwrite any buffer portion backing a Reserve()d (Sub)Region
+//
+// Implementation notes
+//
+// Unlike io.ReadAtLeast, errors from the underlying reader are always made
+// available on NextRegion(…). As with the standard io.* semantics an error can
+// be returned together with a result. One should always check whether the
+// *Region return value is nil first, before processing the error,
+//
+// Changes of the emitter and collector positions are protected by a mutex on
+// the qringbuf object. Any call accessing the buffer state will block until
+// this lock can be obtained. One can also lock the mutex if the collector
+// goroutine needs to be paused for whatever reason.
```

Comment:
Avoid exposing mutexes like this. It makes internal locking patterns a part of the API.